vulnfeeds/cpp/main.go

```diff
@@ -75,84 +76,85 @@ func InScopeGitRepo(repoURL string) bool {
-}
-
-// Either clones the repository or returns the one previously cloned.
-func GetRepo(repoURL string, repobasedir string) (r *git.Repository, e error) {
-	repodir := GetRepoDir(repoURL, repobasedir)
-	// TODO: reorder to open, then clone, once behaviour and failure modes understood
-	Logger.Infof("Cloning %s", repoURL)
-	r, err := git.PlainClone(repodir, false, &git.CloneOptions{
-		URL: repoURL,
-	})
-	if err != nil && err != git.ErrRepositoryAlreadyExists {
-		return nil, err
-	}
-	r, err = git.PlainOpen(repodir)
-	if err != nil {
-		return nil, err
-	}
-	return r, nil
-}
-
-// Examines repos and tries to convert version tags to commits
-func GitVersionToCommit(versions cves.VersionInfo, repos []string, repobasedir string) (v cves.VersionInfo, e error) {
+// Examines repos and tries to convert versions to commits by treating them as Git tags.
+// Takes a CVE string (for logging), cves.VersionInfo with AffectedVersions and
+// no FixCommits and attempts to add FixCommits.
+func GitVersionToCommit(CVE string, versions cves.VersionInfo, repos []string, cache git.RepoTagsCache) (v cves.VersionInfo, e error) {
 	// versions is a VersionInfo with AffectedVersions and no FixCommits
 	// v is a VersionInfo with FixCommits included
 	v = versions
 	for _, repo := range repos {
-		r, err := GetRepo(repo, repobasedir)
+		normalizedTags, err := git.NormalizeRepoTags(repo, cache)
 		if err != nil {
-			Logger.Warnf("Failed to get %s: %v", repo, err)
+			Logger.Warnf("[%s]: Failed to normalize tags for %s: %v", CVE, repo, err)
 			continue
 		}
 		for _, av := range versions.AffectedVersions {
 			// We can't map a non-existent version to a tag and commit, so skip.
-			if av.Fixed == "" {
+			if av.Fixed == "" && av.LastAffected == "" {
 				continue
 			}
-			// TODO: add more sophisticated (fuzzy) version to tag matching.
-			ref, err := r.Tag("v" + av.Fixed)
-			if err != nil {
-				Logger.Warnf("Failed to find a tag for %q in %s: %v", av.Fixed, repo, err)
-				continue
+			if av.Fixed != "" {
+				normalizedFixed, err := cves.NormalizeVersion(av.Fixed)
+				if err != nil {
+					Logger.Warnf("[%s]: Failed to normalize fixed version %q: %v", CVE, av.Fixed, err)
+					continue
+				}
+				// Try a straight out match first.
+				// TODO try fuzzy prefix matches also.
+				normalizedTag, ok := normalizedTags[normalizedFixed]
+				if !ok {
+					Logger.Warnf("[%s]: Failed to find a commit for fixed version %q normalized as %q", CVE, av.Fixed, normalizedFixed)
+				} else {
+					gc := cves.GitCommit{
+						Repo:   repo,
+						Commit: normalizedTag.Commit,
+					}
+					Logger.Infof("[%s]: Successfully derived %+v for fixed version %q", CVE, gc, av.Fixed)
+					v.FixCommits = append(v.FixCommits, gc)
+				}
 			}
-			fc := cves.GitCommit{
-				Repo:   repo,
-				Commit: ref.Hash().String(),
+			if av.LastAffected != "" {
+				normalizedLastAffected, err := cves.NormalizeVersion(av.LastAffected)
```

Comment:
These two branches for "Fixed" and "LastAffected" look **very)** similar. Can we create a helper function instead for these, and parameterize the bits that change?

Reply:
Had a stab at addressing in [020dfb1](https://github.com/google/osv.dev/pull/1105/commits/020dfb1ca92fdc38f5aa34f4c9b847e17682a833)